meson: mempcpy is incorrectly detected as available on macOS

The `mempcpy` function doesn't exist on macOS but Meson thinks it does.
I've produced the issue with Meson 1.3.2 and 1.4.0.
It is likely related to the behavior for checking built-ins in Meson's `compiler.has_function` functionality.

```
...
Checking for function "mempcpy" : YES 
...
[1/155] Compiling C object lib/libcommon.a.p/buffer.c.o
FAILED: lib/libcommon.a.p/buffer.c.o 
clang -Ilib/libcommon.a.p -Ilib -I../src/lib -Iinclude -I../src/include -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c99 -O3 -D_GNU_SOURCE -include /Users/jenkins/workspace/prod-v1/bsr/11724/eaccf/.conan/data/util-linux-libuuid/2.40/_/_/build/6280371def85df4402bdeba403e98e478874576a/build-release/config.h -fno-common -Waddress-of-packed-member -Wembedded-directive -Wextra-semi -Wformat-security -Wimplicit-function-declaration -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wno-missing-field-initializers -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wsign-compare -Wstrict-prototypes -Wtype-limits -Wuninitialized -Wunused-parameter -Wunused-result -Wunused-variable -Wvla -Walloca -isysroot /Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -arch arm64 -MD -MQ lib/libcommon.a.p/buffer.c.o -MF lib/libcommon.a.p/buffer.c.o.d -o lib/libcommon.a.p/buffer.c.o -c ../src/lib/buffer.c
../src/lib/buffer.c:150:13: error: implicit declaration of function 'mempcpy' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        buf->end = mempcpy(buf->end, data, sz);
                   ^
../src/lib/buffer.c:150:11: warning: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
        buf->end = mempcpy(buf->end, data, sz);
                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/meson.build
+++ b/meson.build
@@ -569,7 +569,6 @@ funcs = '''
         llistxattr
         llseek
         newlocale
-        mempcpy
         mkostemp
         move_mount
         mount_setattr
@@ -633,6 +632,9 @@ foreach func: funcs
   # redefined macros from Python.h, which uses this convention.
   conf.set('HAVE_' + func.to_upper(), have ? 1 : false)
 endforeach
+
+have_mempcpy = cc.has_function('mempcpy', prefix: '#include <string.h>', args: '-D_GNU_SOURCE')
+conf.set('HAVE_MEMPCPY', have_mempcpy ? 1 : false)
 
 have = conf.get('HAVE_FUTIMENS') in [1] and conf.get('HAVE_INOTIFY_INIT1') in [1]
 conf.set('AGETTY_RELOAD', have ? 1 : false)

